File: 卡码/代码随想录/1/09-动态规划/Java/dp/src/main/java/com/kurumi/dp/_05_子序列问题/_02_连续子序列/LeetCode_0053.java
//53. 最大子数组和
package com.kurumi.dp._05_子序列问题._02_连续子序列;

public class LeetCode_0053 {

}
class Solution {
  public int maxSubArray(int[] nums) {
    int[] dp=new int[nums.length];
    dp[0]=nums[0];
    int res=nums[0];
    for(int i=1;i<nums.length;i++){
      dp[i]=Math.max(dp[i-1]+nums[i],nums[i]);
      res=dp[i]>res?dp[i]:res;
    }
    return res;
  }
}

File: 卡码/代码随想录/1/08-贪心算法/Go/03-困难/LeetCode_0053.go
// 53. 最大子数组和
package main

import "math"

func maxSubArray(nums []int) int {
	if len(nums)<=1 {
		return len(nums)
	}
	res,cnt:=math.MinInt,0
	for _,num:=range nums {
		cnt+=num
		if cnt>res {
			res=cnt
		}
		if cnt<=0 {
			cnt=0
		}
	}
	return res
}
